pkg/releasing/git.bzl

```diff
@@ -0,0 +1,111 @@
+# Copyright 2021 The Bazel Authors. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""A rule to extract the git changelog."""
+
+def _git_changelog_impl(ctx):
+    """Implements to git_changelog rule."""
+
+    args = ctx.actions.args()
+    tools = []
+
+    toolchain = ctx.toolchains["@rules_pkg//toolchains/git:git_toolchain_type"].git
+    if not toolchain.valid:
+        fail("The git_toolchain is not properly configured: " +
+             toolchain.name)
+    if toolchain.path:
+        args.add("--git_path", toolchain.path)
+    else:
+        executable = toolchain.label.files_to_run.executable
+        tools += [executable]
+        tools += toolchain.label.default_runfiles.files.to_list()
+        args.add("--git_path", executable.path)
+    args.add("--git_root", toolchain.workspace_top)
+    args.add("--from_ref", ctx.attr.from_ref)
+    args.add("--to_ref", ctx.attr.to_ref)
+    args.add("--out", ctx.outputs.out.path)
+    if ctx.attr.verbose:
+        args.add("--verbose")
+
+    # TODO(aiuto): Remove this when we graduate from experimental
+    # Here is the place where we would try to get the absolute path to self.
+    # What we need is a method to return a value of type path, but everything
+    # gives us strings.
+    # print(dir(ctx.file.self))
+    # print(dir(ctx.file.self.root))
+    # print(type(ctx.file.self.path))
+    # print(ctx.file.self.path)
+    # rctx has path(), but ctx does not :-(
+    # print(ctx.path(ctx.attr.self.label).dirname)
+
+    ctx.actions.run(
+        mnemonic = "GitChangelog",
+        executable = ctx.executable._git_changelog,
+        use_default_shell_env = True,
+        arguments = [args],
+        outputs = [ctx.outputs.out],
+        env = {
+            "LANG": "en_US.UTF-8",
+            "LC_CTYPE": "UTF-8",
+            "PYTHONIOENCODING": "UTF-8",
+            "PYTHONUTF8": "1",
+        },
+        execution_requirements = {
+            "local": "1",
+            "no-sandbox": "1",
+            "no-remote": "1",
+        },
+        tools = tools,
+    )
+
+# Define the rule.
+_git_changelog = rule(
```

Comment:
Oh, one more thing: what's the way users are supposed to regenerate logs if something has changed since the last time they ran git?

Reply:
I have no idea at this time. Right now I am just trying to get the toolchain in place and see an initial use of it.   How this works in practice is TBD.
My favorite path is that a human manually does a `bazel run :update_changelog --update` which would generate the change log, then actually update a file in your client.
I don't believe we will ever get to high quality automatic changelog processing and release note extraction.